Remove rustc's notion of "preferred" alignment AKA `__alignof`

workingjubilee opened the pull request:
In PR rust-lang/rust#90877 T-lang decided not to remove `intrinsics::pref_align_of`. However, the intrinsic and its supporting code
1.  is a nightly feature, so can be removed at compiler/libs discretion
2.  requires considerable effort in the compiler to support, as it necessarily complicates every single site reasoning about alignment
3.  has been justified based on relevance to codegen, but it is only a requirement for C++ (not C, not Rust) stack frame layout for AIX[^1], in ways Rust would not consider even with increased C++ interop
4.  is only used by rustc to overalign some globals, not correctness[^2]
5.  can be adequately replaced by other rules for globals, as it mostly affects alignments for a few types under 16 bytes of alignment
6.  has only one clear beneficiary: automating C -> Rust translation for GNU extensions like `__alignof`[^3]
7.  such code was likely intended to be `alignof` or `_Alignof`, because the GNU extension is a "false friend" of the C keyword, which makes the choice to support such a mapping very questionable
8.  makes it easy to do incorrect codegen in the compiler by its mere presence as usual Rust rules of alignment (e.g. `size == align * N`) do not hold with preferred alignment[^4]

Despite an automated translation tool like c2rust using it, we have made multiple attempts to find a crate that actually has been committed to public repositories, like GitHub or crates.io, using such translated code. We have found none. While it is possible someone privately uses this intrinsic, it seems unlikely, and it is behind a feature gate that will warn about using the internal features of rustc.

The implementation is clearly damaging the code quality of the compiler. Thus it is within the compiler team's purview to simply rip it out. If T-lang wishes to have this intrinsic restored for c2rust's benefit, it would have to use a radically different implementation that somehow does not cause internal incorrectness.

Until then, remove the intrinsic and its supporting code, as one tool and an ill-considered GCC extension cannot justify risking correctness.

Because we touch a fair amount of the compiler to change this at all, and unfortunately the duplication of AbiAndPrefAlign is deep-rooted, we keep an "AbiAlign" type which we can wean code off later.

[^1]: https://github.com/rust-lang/rust/issues/91971#issuecomment-2451330704
[^2]: as viewable in the code altered by this PR
[^3]: c2rust: https://github.com/immunant/c2rust/blame/3b1ec86b9b0cf363adfd3178cc45a891a970eef2/c2rust-transpile/src/translator/mod.rs#L3175
[^4]: https://github.com/rust-lang/rustc_codegen_cranelift/issues/1560

<!-- homu-ignore:start -->
r? @bjorn3
<!-- homu-ignore:end -->

workingjubilee committed:
Remove rustc's notion of "preferred" alignment AKA `__alignof`

In PR 90877 T-lang decided not to remove `intrinsics::pref_align_of`.
However, the intrinsic and its supporting code
1.  is a nightly feature, so can be removed at compiler/libs discretion
2.  requires considerable effort in the compiler to support, as it
    necessarily complicates every single site reasoning about alignment
3.  has been justified based on relevance to codegen, but it is only a
    requirement for C++ (not C, not Rust) stack frame layout for AIX,
    in ways Rust would not consider even with increased C++ interop
4.  is only used by rustc to overalign some globals, not correctness
5.  can be adequately replaced by other rules for globals, as it mostly
    affects alignments for a few types under 16 bytes of alignment
6.  has only one clear benefactor: automating C -> Rust translation
    for GNU extensions like `__alignof`
7.  such code was likely intended to be `alignof` or `_Alignof`,
    because the GNU extension is a "false friend" of the C keyword,
    which makes the choice to support such a mapping very questionable
8.  makes it easy to do incorrect codegen in the compiler by its mere
    presence as usual Rust rules of alignment (e.g. `size == align * N`)
    do not hold with preferred alignment

The implementation is clearly damaging the code quality of the compiler.
Thus it is within the compiler team's purview to simply rip it out.
If T-lang wishes to have this intrinsic restored for c2rust's benefit,
it would have to use a radically different implementation that somehow
does not cause internal incorrectness.

Until then, remove the intrinsic and its supporting code, as one tool
and an ill-considered GCC extension cannot justify risking correctness.

Because we touch a fair amount of the compiler to change this at all,
and unfortunately the duplication of AbiAndPrefAlign is deep-rooted,
we keep an "AbiAlign" type which we can wean code off later.

RalfJung commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -3489,15 +3489,6 @@ pub const fn size_of<T>() -> usize;
 #[rustc_intrinsic]
 pub const fn min_align_of<T>() -> usize;
```
Also in the future, we should then rename this to just `align_of`.

RalfJung commented:
I agree with removing the intrinsic. However, the PR description is written in a style that seems somewhat antagonistic to t-lang, which is wholly unnecessary. Quoting some t-lang members (not necessarily speaking for the team) from the prior PR:

> With my lang hat on, I consider intrinsics entirely a compiler-team question so long as they're not otherwise exposed. So removing something unexposed sounds fine to me, FWIW.

> We discussed this both in last week's and today's @rust-lang/lang meeting. In both cases, we came to the conclusion that we're not in a hurry to remove this intrinsic, and it's not obvious that the intrinsic couldn't be stabilized in the future if it makes sense to do so, given that people are using it.
>
> Given that, we don't want to merge this PR at this time, and we don't mind letting this sit on nightly, unless there's some other reason that we should be considering this an urgent issue.

It's not become an "urgent" issue, but the past 3 years have not shown any attempt of this moving any closer to stabilization -- and meanwhile we constantly pay the complexity cost in the compiler. An attempt at exposing this would have to include a clear idea of what "preferred alignment" even *means*, which is a discussion that would be good to have for a clean implementation. IMO that is enough to justify removal.

However, we could help ease the transition by marking it as deprecated for a cycle, pointing to a tracking issue for the removal so that nightly users relying on it have a chance to speak up.

workingjubilee commented:
@RalfJung To be clear, codegen backends have miscompiled Rust code by dint of this intrinsic relying on data that can undermine Rust assumptions: https://github.com/rust-lang/rustc_codegen_cranelift/issues/1560

I am not sure "urgent" is the word but perhaps "should have been fixed yesterday"?

RalfJung commented:
> I am not sure "urgent" is the word but perhaps "should have been fixed yesterday"?

"Should have been fixed some time in the last 10 years" :D (but waiting another few weeks to give people a heads-up is not a big deal)

workingjubilee commented:
I... don't know if it is indeed not a big deal?

I opened this because I remembered that I have to fix it before I do much more meaningful work on layout-related code. We've waited three years for someone to explain why they really need this and nothing has happened.

workingjubilee commented:
I would rather you just close my PR than pretend we're actually helping anyone by implementing a deprecation cycle for a feature no one really wants.

RalfJung commented:
All I did was raise a concern that given how long this intrinsic was around, people might be relying on it, so it might be courteous to give them a heads-up. So far, we have not given a clear signal that this will disappear. OTOH it is guarded by an "internal" feature gate so whoever was using this should be aware they are on thin ice.

If others don't think we need to be this careful, that's fine for me as well, but I think this option should at least be considered.

workingjubilee commented on compiler/rustc_ty_utils/src/layout.rs:
```diff
@@ -526,7 +526,6 @@ fn layout_of_uncached<'tcx>(
                 .map(|mut layout| {
                     // this is similar to how ReprOptions populates its field_shuffle_seed
                     layout.randomization_seed = tcx.def_path_hash(def_id).0.to_smaller_hash();
-                    debug!("coroutine layout ({:?}): {:#?}", ty, layout);
```
Probably not!

workingjubilee commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -863,25 +862,24 @@ impl Align {
 #[cfg_attr(feature = "nightly", derive(HashStable_Generic))]
-pub struct AbiAndPrefAlign {
+pub struct AbiAlign {
     pub abi: Align,
```
Correct.

There's several hundred use sites that will need to be changed and unfortunately it's... challenging... to automatically find and refactor them as there are other cases where we have a field named `abi` :face_with_spiral_eyes:

workingjubilee replied:
I think to help wean code off it, it might be useful for this to impl Deref and DerefMut to the inner Align. What do you think?

workingjubilee commented:
I'm not particularly trying to give T-lang a super-hard time here, but they made a decision, and I don't think this PR makes sense without including a rebuttal of that decision. If anyone should get an earful, it's whoever working on GCC thought it was a good idea to retain `__alignof` instead of harmonizing it with `alignof`.

bors commented:
:umbrella: The latest upstream changes (presumably #142003) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15448311492/job/43483927625) [(plain)](https://github.com/rust-lang/rust/commit/676b32b7ecef26fc14050f343d066fad11b6d645/checks/43483927625/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args intrinsics/intrinsic-alignment.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/intrinsics/intrinsic-alignment.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/intrinsics/intrinsic-alignment/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: unnecessary `unsafe` block
##[warning]  --> /checkout/tests/ui/intrinsics/intrinsic-alignment.rs:33:9
   |
```

</details>

workingjubilee committed:
compiler: add Deref to AbiAlign to ease transition

We will want to remove many cases of `.abi`, including `.abi.thing`,
so this may simplify future PRs and certainly doesn't hurt.

We omit DerefMut because mutation is much rarer and localized.

workingjubilee commented:
rebased because conflict and also "x.py wonky" and also pushed another commit motivated by making the future transition smoother.

bjorn3 reviewed (approved):
Changes LGTM. Do I understand it correctly that this is waiting on a T-lang decision?

workingjubilee commented:
@bjorn3 It is not. I am not asking T-lang if they want to reconsider having an implementation of this feature in the future. I am asserting it is our right to simply remove the entire implementation of a nightly feature.

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout remove-pref-align` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self remove-pref-align --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging library/core/src/intrinsics/mod.rs
Auto-merging compiler/rustc_hir_analysis/src/check/intrinsic.rs
CONFLICT (content): Merge conflict in compiler/rustc_hir_analysis/src/check/intrinsic.rs
Auto-merging compiler/rustc_codegen_ssa/src/mir/intrinsic.rs
Auto-merging compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs
Auto-merging compiler/rustc_abi/src/lib.rs
Auto-merging compiler/rustc_abi/src/layout/ty.rs
Auto-merging compiler/rustc_abi/src/layout.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

workingjubilee commented:
@bors r=bjorn3

bors commented:
:pushpin: Commit ec13ae6475275d82ec48e2ae32e27eaecdfa19cd has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 ec13ae6475275d82ec48e2ae32e27eaecdfa19cd -->
<!-- homu: {"type":"Approved","sha":"ec13ae6475275d82ec48e2ae32e27eaecdfa19cd","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
